#!/bin/bash

#### vlinc1541
python allele.rnaseq.given.windows.py --df ../data/gm12878.4xAligned.out.sorted.minus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/vlinc1541.nochr.sorted.minus.hg38.bed --out_directory ./
python allele.rnaseq.given.windows.py --df ../data/gm12878.4xAligned.out.sorted.plus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/vlinc1541.nochr.sorted.plus.hg38.bed --out_directory ./
python allele.rnaseq.given.windows.py --df ../data/gm12878.5xAligned.out.sorted.minus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/vlinc1541.nochr.sorted.minus.hg38.bed --out_directory ./
python allele.rnaseq.given.windows.py --df ../data/gm12878.5xAligned.out.sorted.plus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/vlinc1541.nochr.sorted.plus.hg38.bed --out_directory ./
python allele.rnaseq.given.windows.py --df ../data/gm12878.rep1Aligned.out.sorted.minus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/vlinc1541.nochr.sorted.minus.hg38.bed --out_directory ./
python allele.rnaseq.given.windows.py --df ../data/gm12878.rep1Aligned.out.sorted.plus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/vlinc1541.nochr.sorted.plus.hg38.bed --out_directory ./
python allele.rnaseq.given.windows.py --df ../data/gm12878.rep2Aligned.out.sorted.minus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/vlinc1541.nochr.sorted.minus.hg38.bed --out_directory ./
python allele.rnaseq.given.windows.py --df ../data/gm12878.rep2Aligned.out.sorted.plus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/vlinc1541.nochr.sorted.plus.hg38.bed --out_directory ./

#### vlinc2149
python allele.rnaseq.given.windows.py --df ../data/gm12878.4xAligned.out.sorted.minus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/vlinc2149.nochr.sorted.minus.hg38.bed --out_directory ./
python allele.rnaseq.given.windows.py --df ../data/gm12878.4xAligned.out.sorted.plus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/vlinc2149.nochr.sorted.plus.hg38.bed --out_directory ./
python allele.rnaseq.given.windows.py --df ../data/gm12878.5xAligned.out.sorted.minus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/vlinc2149.nochr.sorted.minus.hg38.bed --out_directory ./
python allele.rnaseq.given.windows.py --df ../data/gm12878.5xAligned.out.sorted.plus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/vlinc2149.nochr.sorted.plus.hg38.bed --out_directory ./
python allele.rnaseq.given.windows.py --df ../data/gm12878.rep1xAligned.out.sorted.minus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/vlinc2149.nochr.sorted.minus.hg38.bed --out_directory ./
python allele.rnaseq.given.windows.py --df ../data/gm12878.rep1xAligned.out.sorted.plus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/vlinc2149.nochr.sorted.plus.hg38.bed --out_directory ./
python allele.rnaseq.given.windows.py --df ../data/gm12878.rep2xAligned.out.sorted.minus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/vlinc2149.nochr.sorted.minus.hg38.bed --out_directory ./
python allele.rnaseq.given.windows.py --df ../data/gm12878.rep2xAligned.out.sorted.plus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/vlinc2149.nochr.sorted.plus.hg38.bed --out_directory ./

#### whole gene
python allele.rnaseq.given.windows.py --df ../data/gm12878.4xAligned.out.sorted.minus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/ucsc.whole.gene.minus.filtered.hg38.bed --out_directory ./
python allele.rnaseq.given.windows.py --df ../data/gm12878.4xAligned.out.sorted.plus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/ucsc.whole.gene.plus.filtered.hg38.bed --out_directory ./
python allele.rnaseq.given.windows.py --df ../data/gm12878.5xAligned.out.sorted.minus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/ucsc.whole.gene.minus.filtered.hg38.bed --out_directory ./
python allele.rnaseq.given.windows.py --df ../data/gm12878.5xAligned.out.sorted.plus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/ucsc.whole.gene.plus.filtered.hg38.bed --out_directory ./
python allele.rnaseq.given.windows.py --df ../data/gm12878.rep1Aligned.out.sorted.minus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/ucsc.whole.gene.minus.filtered.hg38.bed --out_directory ./
python allele.rnaseq.given.windows.py --df ../data/gm12878.rep1Aligned.out.sorted.plus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/ucsc.whole.gene.plus.filtered.hg38.bed --out_directory ./
python allele.rnaseq.given.windows.py --df ../data/gm12878.rep2Aligned.out.sorted.minus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/ucsc.whole.gene.minus.filtered.hg38.bed --out_directory ./
python allele.rnaseq.given.windows.py --df ../data/gm12878.rep2Aligned.out.sorted.plus.platinum.sites.haplotypes.bed --window_file /Users/heskett/replication_rnaseq/annotation.files/ucsc.whole.gene.plus.filtered.hg38.bed --out_directory ./

#### discovery mode
#### use vlinc.identifier.sh to get windows
